tamr_toolbox/dataset/manage.py

```diff
@@ -0,0 +1,225 @@
+    """
+    dataset_name = dataset.name
+    if dataset.upstream_datasets():
+        raise ValueError(f"{dataset_name} is not a source dataset")
+
+    # Update description and tags
+    dataset_spec = dataset.spec()
+    if description:
+        dataset_spec = dataset_spec.with_description(description)
+    if tags:
+        dataset_spec = dataset_spec.with_tags(tags)
+
+    dataset_spec.put()
+
+    if attributes:
+        # Create attributes from input
+        attribute_specs = _create_specs(
+            attribute_names=attributes, attribute_types=attribute_types
+        )
+
+        # Get current dataset attributes
+        target_dataset_attributes = dataset.attributes
+        target_attribute_dict = {}
+        for attr in target_dataset_attributes.stream():
+            target_attribute_dict[attr.name] = attr
+        primary_keys = dataset.spec().to_dict()["keyAttributeNames"]
+
+        # Update attributes in dataset
+        for attribute in attribute_specs:
+            attr_spec_dict = attribute.to_dict()
+            attribute_name = attr_spec_dict["name"]
+            if attribute_name in primary_keys:
+                continue
+            elif attribute_name in target_attribute_dict.keys():
+                # This attribute already exists, update to new type
+                new_type = AttributeType(attribute.to_dict()["type"])
+                target_attribute = target_attribute_dict[attribute_name]
+                new_attr_spec = target_attribute.spec().with_type(new_type.spec())
+
+                # Remove and add attribute with new type
+                target_dataset_attributes.delete_by_resource_id(target_attribute.resource_id)
+                target_dataset_attributes.create(new_attr_spec.to_dict())
+            else:
+                # This attribute does not already exist, create
+                target_dataset_attributes.create(attr_spec_dict)
+
+        # Remove any attributes from dataset that aren't in the new list of attributes
+        attribute_names = [attr.to_dict()["name"] for attr in attribute_specs]
+        for existing_attribute in target_attribute_dict.keys():
+            if existing_attribute not in attribute_names:
+                target_dataset_attributes.delete_by_resource_id(
+                    target_attribute_dict[existing_attribute].resource_id
+                )
+
+    return dataset
+
+
+def exists(*, target_instance: Client, dataset: str) -> bool:
+    """Check if the dataset exists on target instance
+
+    Args:
+        target_instance: Tamr python client object for the target instance
+        dataset: The dataset name
+
+    Return:
+        True or False for if the dataset exists in target instance
+    """
+
+    try:
+        target_instance.datasets.by_name(dataset)
+    except KeyError:
+        return False
+
+    return True
+
+
+def _create_specs(
+    *, attribute_names: List[str], attribute_types: Union[List[JsonDict], None]
```

Comment:
I think instead of attribute types being a List, it should be a Dict. For example using your example config.
Instead of:  
[
    {"baseType": "STRING"},
    {"baseType": "ARRAY", "innerType": {"baseType": "STRING"}},
    {"baseType": "ARRAY", "innerType": {"baseType": "STRING"}},
    {"baseType": "ARRAY", "innerType": {"baseType": "INT"}},
    {"baseType": "DOUBLE"},
]

it could be: 
[
   "unique_id":  {"baseType": "STRING"},
    "name":  {"baseType": "ARRAY", "innerType": {"baseType": "STRING"}},
    "address":  {"baseType": "ARRAY", "innerType": {"baseType": "STRING"}},
    "user_id": {"baseType": "ARRAY", "innerType": {"baseType": "INT"}},
    "sales": {"baseType": "DOUBLE"},
]

Reply:
I love this idea. It could be more flexible. Any attributes that aren't in the types dict will be defaulted to Array String

Reply:
Thanks! Updated